app/models/container_label_tag_mapping/mapper.rb

```diff
@@ -60,6 +60,7 @@ def self.references_to_tags(tag_references)
     private
 
     def map_label(type, label)
+      type = 'VmAzure' if type == 'Vm' && label[:source] == 'azure'
```

Comment:
this seems to be a bad place to put it, there is only generic code. How are we handling Amazon @cben ?

Reply:
I can put it in the collecting code, but there will be two places at least, and since this method looks like some gateway method for all related incoming data processing, I found it better to put this crutch code here.
Unfortunately all this label tag mapping code looks like bunch of a crutches for a crutches sitting on the crutches. I found it impossible to do a deep refactoring because of backporting needs.

Reply:
We don't yet have 3 All/Vm/VmAzure levels of granularity.  We have just All (`nil`) and Vm, which this change is trying to change.
I agree though that hardcoding here isn't nice.  It'd be very hard to guess this is happening when looking at caller code (it's already confusing that "type" is usually but not always a model name...).
I suggest modifying the signature to take a list of types a mapping can match.  I suggest including `nil` in it, for clarity.  It may be easier to add new method instead of modifying existing method across repos:
```
def map_types(types, label)  # naming is hard
  types.collect_concat do |type|
    map_name_type_value(label[:name], type, label[:value])
  end
end

# TODO: use map_types() everywhere, remove this when unused
def map_label(type, label)
  map_types([type, nil], label)
end
```

Then Azure refresh would call `map_types(['VmAzure', 'Vm', nil], ...)`, and Amazon refresh  `map_types(['VmAmazon', 'Vm', nil], ...)`.

Reply:
I agree!  I just started writing that this is all a crutch to encode 2 fields — provider and type — in one column...
So, do you need additional granularity backported?

I strongly feel extra granularity is not crucial feature.  So in existing releases, one would be able to map "Vm", of both providers — what's the harm in that?  (IMHO one doesn't even need that, "\<All>" mappings are good enough in practice!)

Let's stop adding crutches "because backporting".  That's how we (well, I :flushed:) built this mess.  Let's do minimal workable backporting, and instead refactor master.

Reply:
It was your proposal to distinct providers using `Vm` for Amazon and `VmAzure` for Azure, remember?
I'm going to remove that line in favour of specific change in the Azure collecting code.

Reply:
## ignore everything I said above
I misread the PR intention :-(  After a chat, turns out we both are not interested in adding 3rd level of granularity.

OK, so IIUC the goal here is just `Vm` = amazon vm, `AzureVm` = azure vm, done this way so it can be backported without any for compatibility.  LGTM, though consider just passing "AzureVm" from azure refresh?

Reply:
> should be the right STI which is ...Providers::Azure::Vm

It could be, not a must.  these "type" strings are not really model names, just arbitrary string that should match between mapping created by UI and refresh calling map_labels().
(They started as base model names — "ContainerFoo" — which is not not always *the* STI type of the actual record being tagged.  And I think amazon already departed from that to distinguish vm vs image?)
So you could choose `...Providers::Azure::Vm` (or anything you want), not sure that would be a win or more misleading, I'm good with anything.

Reply:
that line was removed in favour of https://github.com/ManageIQ/manageiq-providers-azure/pull/231/files#diff-1f54967e2f7cd1793ef76031aa337270R287